### PR TITLE
Handle left anchored search for json query dsl

### LINKED ISSUE #4138: Left anchored search does not work when json query dsl is on
### Expected behavior
When the JSON query DSL is on, left anchored searching results from the regular search page are significantly different from how they are with it off.

### Actual behavior
- Get much much fewer results with JSON query DSL turned on

### Steps to replicate
1. Spin up development environment as admin `ORANGELIGHT_ADMIN_NETIDS="your_netid" bundle exec rails s`
2. Go to http://localhost:3000/features
3. Make sure the JSON query DSL is turned *off*
4. Go to `/`
5. Select "Title starts with" from the search dropdown
6. Put "the" in the search box
7. Click enter
8. Note number of search results (for me it is 26)
9. Go back to http://localhost:3000/features and turn the JSON query DSL *on*
10. Repeat steps 4-8, noting that there are significantly fewer results for me it is 1

### Impact of this bug
We cannot upgrade to Blacklight 8 until this bug is addressed

### Implementation notes, if any
This bug is reflected in the following test files when you turn on the JSON query DSL on in test by default:
- spec/features/search_input_facets_spec.rb
- spec/requests/request_spec.rb:175
- spec/requests/request_spec.rb:199
- spec/requests/request_spec.rb:224
- spec/requests/request_spec.rb:237

Sorry for the wall of code. Below are the two resulting Solr queries:

```ruby
# json query dsl off, left anchored query
# Solr query: get select
# 26 search results
{ "qt" => nil,
  "qf" => "${left_anchor_qf}",
  "pf" => "${left_anchor_pf}",
  "q" => "the*",
  "facet" => true,
  "facet.field" => ["access_facet",
                    "location",
                    "format",
                    "{!ex=pub_date_start_sort_single}pub_date_start_sort",
                    "language_facet",
                    "subject_topic_facet",
                    "genre_facet",
                    "subject_era_facet",
                    "instrumentation_facet",
                    "publication_place_facet",
                    "sudoc_facet",
                    "advanced_location_s",
                    "subject_facet"],
  "f.access_facet.facet.sort" => "index",
  "f.location.facet.sort" => "index",
  "f.location.facet.mincount" => 1,
  "f.location.facet.limit" => 21,
  "f.format.facet.sort" => "index",
  "f.format.facet.mincount" => 1,
  "f.format.facet.limit" => 16,
  "f.language_facet.facet.limit" => 11,
  "f.subject_topic_facet.facet.limit" => 11,
  "f.genre_facet.facet.limit" => 11,
  "f.subject_era_facet.facet.limit" => 11,
  "facet.query" => ["cataloged_tdt:[NOW/DAY-7DAYS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-14DAYS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-21DAYS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-1MONTH TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-2MONTHS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-3MONTHS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-6MONTHS TO NOW/DAY+1DAY]",
                    "pub_date_start_sort:[1100 TO 1199]",
                    "pub_date_start_sort:[1200 TO 1299]",
                    "pub_date_start_sort:[1300 TO 1399]",
                    "pub_date_start_sort:[1400 TO 1499]",
                    "pub_date_start_sort:[1500 TO 1599]",
                    "pub_date_start_sort:[1600 TO 1699]",
                    "pub_date_start_sort:[1700 TO 1799]",
                    "pub_date_start_sort:[1800 TO 1899]",
                    "pub_date_start_sort:[1900 TO 1999]",
                    "pub_date_start_sort:[2000 TO 2025]"],
  "f.instrumentation_facet.facet.limit" => 11,
  "f.publication_place_facet.facet.limit" => 11,
  "facet.pivot" => ["lc_1letter_facet,lc_rest_facet"],
  "f.sudoc_facet.facet.sort" => "index",
  "f.sudoc_facet.facet.limit" => 11,
  "f.advanced_location_s.facet.sort" => "alpha",
  "rows" => 20,
  "sort" => "score desc, pub_date_start_sort desc, title_sort asc",
  "stats" => "true",
  "stats.field" => ["pub_date_start_sort"] }

# json query dsl on
# Solr query: get advanced
# One search result - "The Times higher education supplement"
{ "qt" => nil,
  "qf" => "${left_anchor_qf}",
  "pf" => "${left_anchor_pf}",
  "json" => { "query" => { "bool" => { "must" => [{ edismax: { qf: "${left_anchor_qf}", pf: "${left_anchor_pf}", query: "the" } }] } } },
  "facet" => true,
  "facet.field" => ["access_facet",
                    "location",
                    "format",
                    "{!ex=pub_date_start_sort_single}pub_date_start_sort",
                    "language_facet",
                    "subject_topic_facet",
                    "genre_facet",
                    "subject_era_facet",
                    "instrumentation_facet",
                    "publication_place_facet",
                    "sudoc_facet",
                    "advanced_location_s",
                    "subject_facet"],
  "f.access_facet.facet.sort" => "index",
  "f.location.facet.sort" => "index",
  "f.location.facet.mincount" => 1,
  "f.location.facet.limit" => 21,
  "f.format.facet.sort" => "index",
  "f.format.facet.mincount" => 1,
  "f.format.facet.limit" => 16,
  "f.language_facet.facet.limit" => 11,
  "f.subject_topic_facet.facet.limit" => 11,
  "f.genre_facet.facet.limit" => 11,
  "f.subject_era_facet.facet.limit" => 11,
  "facet.query" => ["cataloged_tdt:[NOW/DAY-7DAYS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-14DAYS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-21DAYS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-1MONTH TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-2MONTHS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-3MONTHS TO NOW/DAY+1DAY]",
                    "cataloged_tdt:[NOW/DAY-6MONTHS TO NOW/DAY+1DAY]",
                    "pub_date_start_sort:[1100 TO 1199]",
                    "pub_date_start_sort:[1200 TO 1299]",
                    "pub_date_start_sort:[1300 TO 1399]",
                    "pub_date_start_sort:[1400 TO 1499]",
                    "pub_date_start_sort:[1500 TO 1599]",
                    "pub_date_start_sort:[1600 TO 1699]",
                    "pub_date_start_sort:[1700 TO 1799]",
                    "pub_date_start_sort:[1800 TO 1899]",
                    "pub_date_start_sort:[1900 TO 1999]",
                    "pub_date_start_sort:[2000 TO 2025]"],
  "f.instrumentation_facet.facet.limit" => 11,
  "f.publication_place_facet.facet.limit" => 11,
  "facet.pivot" => ["lc_1letter_facet,lc_rest_facet"],
  "f.sudoc_facet.facet.sort" => "index",
  "f.sudoc_facet.facet.limit" => 11,
  "f.advanced_location_s.facet.sort" => "alpha",
  "rows" => 20,
  "sort" => "score desc, pub_date_start_sort desc, title_sort asc",
  "stats" => "true",
  "stats.field" => ["pub_date_start_sort"],
  "q" => nil }
```

### PR DESCRIPTION
**Pull request by maxkadel**
- Separate out escaping whitespace and adding the wildcard at the end
- factor out finding the query 
- factor out testing out whether it is a left anchor search

Closes #4138

### CODE REVIEW

File: app/helpers/blacklight_helper.rb [hunk 1]
```diff
@@ -19,18 +19,51 @@ def wildcard_char_strip(solr_parameters)
...
 
+  def prepare_left_anchor_search(solr_parameters)
+    return unless left_anchor_search?(solr_parameters)
+    left_anchor_escape_whitespace(solr_parameters)
+    left_anchor_add_wildcard(solr_parameters)
+  end
+
+  def left_anchor_query(solr_parameters)
+    if Flipflop.json_query_dsl?
+      solr_parameters.dig('json', 'query', 'bool', 'must', 0, :edismax, :query)
```
Comment (sandbergja) [inline app/helpers/blacklight_helper.rb:30]: I'm wondering about the `0` here.  Will this only work for the first search term (i.e. if a user did a left-anchor search in the second input in the advanced search form, would it still work)?

In the `SearchBuilder`, I noticed we [do a map](https://github.com/pulibrary/orangelight/blob/4427e81247c62bf51a9db29eeedd958c44cec966/app/models/search_builder.rb#L24-L26) to catch everything in the `must` array.

File: app/helpers/blacklight_helper.rb [hunk 2]
```diff
@@ -19,18 +19,51 @@ def wildcard_char_strip(solr_parameters)
...
+  def left_anchor_search?(solr_parameters)
+    query = left_anchor_query(solr_parameters)
+    return false unless (solr_parameters[:qf] == '${left_anchor_qf}' && query) || (solr_parameters.dig('json', 'query', 'bool', 'must', 0, :edismax, :qf) == '${left_anchor_qf}' && query)
+    true
+  end
+
   # Escape all whitespace characters within Solr queries specifying left anchor query facets
-  # Ends all left-anchor searches with wildcards for matches that begin with search string
   # @param solr_parameters [Blacklight::Solr::Request] the parameters for the Solr query
   def left_anchor_escape_whitespace(solr_parameters)
```
Comment (sandbergja) [inline app/helpers/blacklight_helper.rb:44]: This is so much nicer!

File: app/helpers/blacklight_helper.rb (hunk 1 above)
Comment (maxkadel) [reply app/helpers/blacklight_helper.rb:30]: Ah, I suspect you're right, thank you!

**Comment by maxkadel** [comment]
Superseded by #4168

**Pull request closed without merging**